### PR TITLE
Feature/read nzbpassword from filename

### PR DESCRIPTION
**Pull request by Jhonny0007**
## Description

This Change adds a function to the nzbfile class to derive unpack password from the filename if it is encoded in {{Brackets}}.

### CODE REVIEW

**Comment by dnzbk** [comment]
I have to reject your changes as they solve a very specific issue.
And the `ReadPasswordFromFilename()` member function won't work correctly if the password starts with `{{` or ends with `}}`.

I'd suggest you to use the [Scan](https://nzbget.com/documentation/scan-scripts/) extensions to achieve the behavior you need.

**Comment by luckedea** [comment]
@Jhonny0007 NZBGet already has a function to read nzbpassword from the file. You suggest a new feature to read from filename - can you expand on what you based it? Are there nzb files like this?

**Comment by Jhonny0007** [comment]
Hi, about 95% of the nzbs that i encounter have the unpack password encoded in the filename, most sites (that i use) dont use the "meta type=password" tag. There is even a plugin for this (https://forum.nzbget.net/viewtopic.php?f=8&t=915&sid=589c3d9766079788998a43ba4bdd9720) that for some reason didnt make it to the official plugin repository. And with the use of kubernetes injecting costume scripts reliably is a pain if you also want to retain the script dir as read/write. So i wanted to fix this issue directly at the source instead of using workarounds that decrease the general usability of nzbget.

File: daemon/queue/NzbFile.cpp [hunk 1]
```diff
@@ -391,15 +391,31 @@ void NzbFile::ProcessFiles()
 
 	if (m_password)
 	{
-		ReadPassword();
+		ReadPasswordFromNzb();
+	}
+	ReadPasswordFromFilename();
```
Comment (dnzbk) [inline daemon/queue/NzbFile.cpp:396]: If `ReadPasswordFromNzb` finds the password do we need to look for another password in the NZB filename?
Shouldn't the logic be like
```
if (m_password)
{
   ReadPasswordFromNzb();
}
else
{
   ReadPasswordFromFilename();
}
```

**Comment by dnzbk** [comment]
@Jhonny0007 

Got it, thanks.

Could you add a couple of unit tests just to be sure?
You can use [NzbFileTest](https://github.com/nzbgetcom/nzbget/blob/develop/tests/queue/NzbFileTest.cpp) module and [testdata](https://github.com/nzbgetcom/nzbget/tree/develop/tests/testdata).

File: daemon/queue/NzbFile.cpp (hunk 1 above)
Comment (Jhonny0007) [reply daemon/queue/NzbFile.cpp]: I have updated the i have updated the logic in the pull request to reflect that change. I wasnt entirely sure what the possible states of the var m_password is, or how the rest of the program responds to it. So far my manual test work as well, so if you think that logic is better, lets go with it.

**Comment by Jhonny0007** [comment]
Sure i will add unit-tests for this feature, just give me some time to figure those out.

**Comment by dnzbk** [comment]
@Jhonny0007 
Sorry to bother you but I noticed in your recent commits that you're just pulling the latest changes from the develop branch - perhaps you've already finished working on your branch? 
If so then just ping me and I'll start reviewing and testing it.

**Comment by Jhonny0007** [comment]
Hi, ye,s, im done with with the code changes, if you guys dont have any objections, or cirtiques.

File: daemon/queue/NzbFile.cpp [hunk 2]
```diff
@@ -391,15 +391,32 @@ void NzbFile::ProcessFiles()
 
 	if (m_password)
 	{
-		ReadPassword();
+		ReadPasswordFromNzb();
+	}else{
```
Comment (dnzbk) [inline daemon/queue/NzbFile.cpp:395]: Please, move 'else' and '{' to the next and separate lines to be consistent with the code style.

File: tests/queue/NzbFileTest.cpp [hunk 3]
```diff
@@ -63,6 +64,16 @@ void TestNzb(std::string testFilename)
...
 		BOOST_CHECK(std::string(fileInfo->GetFilename()) == std::string(buffer));
+		memcpy(lastBuffer, buffer, sizeof(buffer));
+	}
+
+	while (fgets(buffer, sizeof(buffer), infofile) && *buffer == '#') ;
+
+	if(strcmp(lastBuffer, buffer) == 0)
+	{
+		BOOST_CHECK(nzbFile.GetPassword() == nullptr);
+	}else{
```
Comment (dnzbk) [inline tests/queue/NzbFileTest.cpp:75]: Same here.

Comment (Jhonny0007) [reply tests/queue/NzbFileTest.cpp]: should be done. anything else?

**Review by dnzbk — APPROVED** [review]
Thank you!

**Pull request merged**